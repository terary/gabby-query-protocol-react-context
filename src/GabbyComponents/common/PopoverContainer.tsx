/* eslint-disable import/prefer-default-export */
/*
To use:
   const parentRef = React.useRef<HTMLElement>(null);
   const [isOpen, setIsOpen] = React.useState()
   ...
   <div ref={parentRef}>
      any stuff goes here
      <button onClick={()=>{setIsOpen(!isOpen)}}>
        toggle open/close 
      </button>        
   </div>
    <PopoverContainer
      children={<SomeComponent />},
      isOpen={isPopOverOpen}
      minWidth="500px",
      onClickAway={()=>{setIsOpen(false)}},
      parentEl={stackRef.current}
      width="500px",
    />
*/

import React, { ReactNode } from "react";
import Popover from "@mui/material/Popover";
import Box from "@mui/material/Box";

interface Props {
  children: ReactNode;
  isOpen: boolean;
  minWidth?: number | string;
  onClickAway: () => void;
  parentEl: HTMLElement | null;
  width?: number | string;
}

export function PopoverContainer({
  children,
  isOpen,
  minWidth,
  onClickAway,
  parentEl,
  width,
}: Props): JSX.Element {
  return (
    <Popover
      // TODO *tmc* this should use theme at least for color, black is the wrong color.
      PaperProps={{ sx: { border: "1px black solid" } }}
      sx={{ width }}
      open={isOpen}
      anchorEl={parentEl}
      onClose={onClickAway}
      transformOrigin={{ horizontal: -10, vertical: -35 }}
      transitionDuration={500}
    >
      <Box sx={{ m: "10px", minWidth }}>{children}</Box>
    </Popover>
  );
}

PopoverContainer.defaultProps = {
  width: "50%",
  minWidth: "250px",
  layout: "3row",
};
