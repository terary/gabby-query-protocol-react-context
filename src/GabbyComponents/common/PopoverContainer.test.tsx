import React, { useRef, useState } from "react";
import renderer from "react-test-renderer";
import {
  act,
  render,
  RenderResult,
  cleanup,
  fireEvent,
  screen,
  waitFor,
} from "@testing-library/react";

import { toMatchDiffSnapshot } from "snapshot-diff";
import { PopoverContainer } from "./PopoverContainer";
import userEvent from "@testing-library/user-event";

expect.extend({ toMatchDiffSnapshot });

type ChildComponentProps = {
  isOpen: boolean;
  onFinish: (output: string) => void;
  onCancel: () => void;
};
const ChildComponent = ({ isOpen, onCancel, onFinish }: ChildComponentProps) => {
  return (
    <div>
      {isOpen && (
        <div>
          <span>Child is Open</span>
          <button
            data-testid="finish-button"
            onClick={() => {
              onFinish("finished called");
            }}
          >
            Finish
          </button>
          <button data-testid="cancel-button" onClick={onCancel}>
            Cancel
          </button>
          <br />
        </div>
      )}
      {!isOpen && (
        <div>
          <span>This will never be seen</span>
          <br />
        </div>
      )}
    </div>
  );
};

type PopoverParentProps = {
  onFinishMock?: (output: string) => void;
  onCancelMock?: () => void;
  clickAwayMock?: () => void;
};
const PopoverParent = ({ clickAwayMock, onCancelMock, onFinishMock }: PopoverParentProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const handleButtonClick = () => {
    setIsOpen(false);
  };
  const parentRef = useRef(null);

  const handleClickAway = () => {
    setIsOpen(false);
    if (clickAwayMock) clickAwayMock();
  };

  return (
    <div ref={parentRef} data-testid="parent-outer-div">
      <button data-testid="parent-open-button" onClick={() => setIsOpen(!isOpen)}>
        Open Child
      </button>
      <button onClick={() => console.log("do nothing")} data-testid="click-away-button">
        Click Away Button
      </button>
      <div data-testid="click-away-area" style={{ height: "50px", width: "50px" }}>
        Click-away area
      </div>
      {isOpen && (
        <PopoverContainer
          children={
            <ChildComponent
              isOpen={isOpen}
              onFinish={onFinishMock || handleButtonClick}
              onCancel={onCancelMock || handleButtonClick}
            />
          }
          isOpen={isOpen}
          onClickAway={handleClickAway}
          parentEl={parentRef.current}
        />
      )}
    </div>
  );
};

describe("Popover Container", () => {
  describe("Snapshots should be unchanged", () => {
    it("Should render component within popover container", () => {
      const tree = renderer.create(<PopoverParent />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  }); // snapshot
  describe("isOpen state", () => {
    beforeEach(cleanup);
    it("Should open/closed depending on parent", async () => {
      // set-up
      const renderedComponent = render(<PopoverParent />);
      const parentOpenButton = await renderedComponent.findByTestId("parent-open-button");

      // pre-conditions
      expect(childIsClosed(renderedComponent)).toBe(true);

      // exercise 1, open child
      // open popover with child
      userEvent.click(parentOpenButton);

      // post-conditions 1
      expect(childIsOpen(renderedComponent)).toBe(true);

      // exercise 2, close child
      // close popover with child
      userEvent.click(parentOpenButton);

      // post-conditions 2
      expect(childIsClosed(renderedComponent)).toBe(true);
    }); // it("Should change open/closed state by setting 'isOpen'",

    it("Should render support children with click handler", async () => {
      const cancelMock = jest.fn();
      const finishMock = jest.fn();
      const renderedComponent = render(
        <PopoverParent onCancelMock={cancelMock} onFinishMock={finishMock} />
      );
      const parentOpenButton = await renderedComponent.findByTestId("parent-open-button");

      userEvent.click(parentOpenButton);

      const childFinishButton = renderedComponent.getByTestId("finish-button");
      const childCancelButton = renderedComponent.getByTestId("cancel-button");

      userEvent.click(childFinishButton);
      userEvent.click(childCancelButton);

      expect(cancelMock).toHaveBeenCalled();
      expect(finishMock).toHaveBeenCalledWith("finished called");
    });

    it.skip("Should accept function for click-away event", async () => {
      // clickAway seem to work fine in dev. Testing, not so much.
      //  Can't seem to test the click-away or mouseLeave events.

      const clickAwayMock = jest.fn(() => {
        console.log("This is click-away");
      });
      const renderedComponent = render(<PopoverParent clickAwayMock={clickAwayMock} />);
      const clickAwayArea = await renderedComponent.findByTestId("click-away-area");
      const clickAwayButton = await renderedComponent.findByTestId("click-away-button");
      const parentOpenButton = await renderedComponent.findByTestId("parent-open-button");

      act(() => {
        userEvent.click(parentOpenButton);
      });
      expect(childIsOpen(renderedComponent)).toBe(true);

      act(() => {
        // click away seem to work but can't seem to get it
        // to work in testing.
        // userEvent.click(clickAwayButton);
        // fireEvent.mouseDown(clickAwayArea);
        // fireEvent.mouseLeave(clickAwayArea);
        userEvent.click(parentOpenButton);
      });

      expect(childIsClosed(renderedComponent)).toBe(true);

      // await waitFor(() => {
      //   const childFinishButton = screen.getByTestId("finish-button");
      //   expect(childFinishButton).not.toBeInTheDocument();
      // });
    });
  }); // describe("isOpen state", () => {
});

const childIsClosed = (renderedComponent: RenderResult) => {
  const childFinishButton = renderedComponent.queryByTestId("finish-button");
  const childCancelButton = renderedComponent.queryByTestId("cancel-button");
  const parentChildOpenMessage = renderedComponent.queryByText(/Child is Open/);

  expect(childFinishButton).not.toBeInTheDocument();
  expect(childCancelButton).not.toBeInTheDocument();
  expect(parentChildOpenMessage).not.toBeInTheDocument();

  return true;
};
const childIsOpen = (renderedComponent: RenderResult) => {
  const childFinishButton = renderedComponent.queryByTestId("finish-button");
  const childCancelButton = renderedComponent.queryByTestId("cancel-button");
  const parentChildOpenMessage = renderedComponent.queryByText(/Child is Open/);

  expect(childFinishButton).toBeInTheDocument();
  expect(childCancelButton).toBeInTheDocument();
  expect(parentChildOpenMessage).toBeInTheDocument();

  return true;
};
