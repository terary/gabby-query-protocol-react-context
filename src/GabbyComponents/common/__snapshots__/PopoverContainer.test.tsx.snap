// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Container Snapshots should be unchanged Should render component within popover container 1`] = `
<div
  data-testid="parent-outer-div"
>
  <button
    data-testid="parent-open-button"
    onClick={[Function]}
  >
    Open Child
  </button>
  <button
    data-testid="click-away-button"
    onClick={[Function]}
  >
    Click Away Button
  </button>
  <div
    data-testid="click-away-area"
    style={
      Object {
        "height": "50px",
        "width": "50px",
      }
    }
  >
    Click-away area
  </div>
</div>
`;
